File: lib/golden_record/substrate.ex
defmodule GoldenRecord.Substrate do
  alias GoldenRecord.{Events, Codes}

  alias Events.ClaimAsserted

  # Every ingested code is canonicalized here so equivalent representations (EAN-13 vs GTIN-14,
  # UPC vs its EAN-13 form, a GTIN-8 vs its zero-padded width) collapse to one identity.
  #
  # :member_of is the legacy spelling of an :edge (gr-xde): the constructor still accepts it,
  # but the LOG holds the generalized edge — one relationship representation, one traversal.
  def claim(source, :member_of, %{member_code: m, collection: c}, valid_from, recorded_at),
    do: claim(source, :edge, %{from: m, relation: :member_of, to: c}, valid_from, recorded_at)

  def claim(source, kind, data, valid_from, recorded_at),
    do: %ClaimAsserted{
      source: source,
      kind: kind,
      data: normalize(kind, data),
      valid_from: valid_from,
      recorded_at: recorded_at
    }

  defp normalize(:identity, %{codes: codes} = d), do: %{d | codes: Enum.map(codes, &Codes.canonicalize/1)}

  defp normalize(:identity_evidence, %{left: left, right: right} = d),
    do: %{d | left: Codes.canonicalize(left), right: Codes.canonicalize(right)}

  defp normalize(:grouping, %{code: c} = d), do: %{d | code: Codes.canonicalize(c)}
  defp normalize(:attribute, %{code: c} = d), do: %{d | code: Codes.canonicalize(c)}
  defp normalize(:media, %{target: t} = d), do: %{d | target: Codes.canonicalize(t)}

  # Both edge endpoints canonicalize, so an edge addressed by EAN-13 matches a cluster holding
  # the GTIN-14 form. (:member_of claims no longer reach here — claim/5 lowers them to :edge —
  # but a previously persisted log may still carry them, so the clause stays foldable.)
  defp normalize(:edge, %{from: f, to: t} = d),
    do: %{d | from: Codes.canonicalize(f), to: Codes.canonicalize(t)}

  defp normalize(:member_of, %{member_code: m, collection: c} = d),
    do: %{d | member_code: Codes.canonicalize(m), collection: Codes.canonicalize(c)}

  defp normalize(_kind, d), do: d

  # Public (@doc false) so the API's fold-state can maintain the current view INCREMENTALLY —
  # one Map.put per claim instead of re-grouping the whole log per projection.
  @doc false
  def slot(%ClaimAsserted{source: s, record_ref: ref} = claim) when not is_nil(ref),
    do: {s, :record, ref, local_slot(claim)}

  def slot(%ClaimAsserted{source: s, kind: :identity, data: %{ref: r}}), do: {s, :identity, r}

  def slot(%ClaimAsserted{source: s, kind: :identity_evidence, data: %{left: left, right: right}}),
    do: {s, :identity_evidence, left, right}

  def slot(%ClaimAsserted{source: s, kind: :grouping, data: %{code: c}}), do: {s, :grouping, c}
  def slot(%ClaimAsserted{source: s, kind: :attribute, data: %{code: c, field: f}}), do: {s, :attr, c, f}
  def slot(%ClaimAsserted{source: s, kind: :media, data: %{asset: a, target: t}}), do: {s, :media, a, t}

  def slot(%ClaimAsserted{source: s, kind: :edge, data: %{from: f, relation: r, to: t}}),
    do: {s, :edge, f, r, t}

  def slot(%ClaimAsserted{source: s, kind: :member_of, data: %{member_code: m, collection: c}}),
    do: {s, :member_of, m, c}

  @doc false
  def local_slot(%ClaimAsserted{kind: :identity}), do: :identity

  def local_slot(%ClaimAsserted{kind: :identity_evidence, data: %{left: left, right: right}}),
    do: {:identity_evidence, left, right}

  def local_slot(%ClaimAsserted{kind: :grouping, data: %{code: c}}), do: {:grouping, c}
  def local_slot(%ClaimAsserted{kind: :attribute, data: %{code: c, field: f}}), do: {:attr, c, f}
  def local_slot(%ClaimAsserted{kind: :media, data: %{asset: a, target: t}}), do: {:media, a, t}

  def local_slot(%ClaimAsserted{kind: :edge, data: %{from: f, relation: r, to: t}}),
    do: {:edge, f, r, t}

  def local_slot(%ClaimAsserted{kind: :member_of, data: %{member_code: m, collection: c}}),
    do: {:member_of, m, c}

  def current(claims) do
    claims
    |> Enum.group_by(&slot/1)
    |> Enum.map(fn {_slot, cs} -> Enum.max_by(cs, & &1.order) end)
  end
end

File: php/src/Substrate.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * Claim construction + the current-view fold — ported from `Substrate` in lib/golden_record_core.ex.
 *
 * Every ingested code is canonicalized here so equivalent representations collapse. `member_of` is
 * the legacy spelling of an `edge`: the constructor lowers it, so the log holds the generalized
 * edge. `current/1` keeps only the latest claim per slot (a slot = the dimension a claim addresses).
 */
final class Substrate
{
    /**
     * Build a {@see Claim}. `member_of` data ({member_code, collection}) lowers to an `edge`.
     *
     * @param array<string,mixed> $data
     */
    public static function claim(?string $source, string $kind, array $data, mixed $validFrom, mixed $recordedAt, mixed $validTo = null, mixed $by = null): Claim
    {
        if ($kind === 'member_of' && isset($data['member_code'], $data['collection'])) {
            return self::claim(
                $source,
                'edge',
                ['from' => $data['member_code'], 'relation' => 'member_of', 'to' => $data['collection']],
                $validFrom,
                $recordedAt,
                $validTo,
                $by
            );
        }

        return Events::claimAsserted($source, $kind, self::normalize($kind, $data), $validFrom, $recordedAt, null, $validTo, $by);
    }

    /**
     * @param array<string,mixed> $data
     * @return array<string,mixed>
     */
    private static function normalize(string $kind, array $data): array
    {
        switch ($kind) {
            case 'identity':
                if (isset($data['codes'])) {
                    $data['codes'] = array_map(Codes::canonicalize(...), $data['codes']);
                }

                return $data;
            case 'identity_evidence':
                if (isset($data['left'], $data['right'])) {
                    $data['left'] = Codes::canonicalize($data['left']);
                    $data['right'] = Codes::canonicalize($data['right']);
                }

                return $data;
            case 'grouping':
            case 'attribute':
                if (isset($data['code'])) {
                    $data['code'] = Codes::canonicalize($data['code']);
                }

                return $data;
            case 'media':
                if (isset($data['target'])) {
                    $data['target'] = Codes::canonicalize($data['target']);
                }

                return $data;
            case 'edge':
                if (isset($data['from'], $data['to'])) {
                    // Elixir canonicalizes BOTH endpoints. For a member_of edge the `to` is a
                    // {collection, member} tuple — Codes::canonicalize just trims its value (the
                    // collection scheme is not GTIN/padded), which is exactly what we want: e.g. a
                    // tab-only member trims to "".
                    $data['from'] = Codes::canonicalize($data['from']);
                    $data['to'] = Codes::canonicalize($data['to']);
                }

                return $data;
            case 'member_of':
                if (isset($data['member_code'], $data['collection'])) {
                    $data['member_code'] = Codes::canonicalize($data['member_code']);
                    $data['collection'] = Codes::canonicalize($data['collection']);
                }

                return $data;
            default:
                return $data;
        }
    }

    /**
     * The slot a claim addresses — its dedup key in `current/1`. Returned as a flat list whose
     * first element is the kind tag, mirroring the Elixir tuple shapes.
     *
     * @return list<mixed>
     */
    public static function slot(Claim $claim): array
    {
        $s = $claim->source;
        $d = $claim->data;

        return match ($claim->kind) {
            'identity' => [$s, 'identity', $d['ref']],
            'identity_evidence' => [$s, 'identity_evidence', $d['left'], $d['right']],
            'grouping' => [$s, 'grouping', $d['code']],
            'attribute' => [$s, 'attr', $d['code'], $d['field']],
            'media' => [$s, 'media', $d['asset'], $d['target']],
            'edge' => [$s, 'edge', $d['from'], $d['relation'], $d['to']],
            'member_of' => [$s, 'member_of', $d['member_code'], $d['collection']],
            default => [$s, $claim->kind],
        };
    }

    /**
     * Collapse the claim log to the latest claim per slot (highest `order` wins). A slot whose
     * latest claim is a {@see retracted} marker has no current claim at all and is dropped.
     *
     * @param list<Claim> $claims
     * @return list<Claim>
     */
    public static function current(array $claims): array
    {
        /** @var array<string, Claim> $bySlot */
        $bySlot = [];
        foreach ($claims as $c) {
            $key = self::slotKey(self::slot($c));
            if (!isset($bySlot[$key]) || ($c->order ?? 0) > ($bySlot[$key]->order ?? 0)) {
                $bySlot[$key] = $c;
            }
        }

        return array_values(array_filter($bySlot, static fn (Claim $c): bool => !self::retracted($c)));
    }

    /**
     * An edge retraction marker (gh-131): the live writer appends the omitted edge again with
     * `data.retracted = true`, so the log keeps the audit trail while the current view drops the
     * slot. The marker is an ordinary claim — it wins its slot by `order`, and a later re-assertion
     * of the same edge (without the flag) wins it back. Attributes need no marker: a `null` value
     * is already how a backfilled null-set reads, so live omission retracts them the same way.
     */
    public static function retracted(Claim $c): bool
    {
        return $c->kind === 'edge' && ($c->data['retracted'] ?? false) === true;
    }

    /**
     * A stable string key for a slot list (codes are [scheme,value] pairs, collections likewise).
     *
     * @param list<mixed> $slot
     */
    public static function slotKey(array $slot): string
    {
        return implode("\x1e", array_map(static function (mixed $part): string {
            if (is_array($part)) {
                return implode("\x1f", array_map(static fn ($x): string => (string) $x, $part));
            }

            return is_bool($part) ? ($part ? '1' : '0') : (string) $part;
        }, $slot));
    }
}
